nwbinspector/checks/check_nwb_containers.py

```diff
@@ -0,0 +1,22 @@
+"""Authors: Cody Baker and Ben Dichter."""
+import pynwb
+import h5py
+
+from ..utils import nwbinspector_check
+
+
+@nwbinspector_check(severity=1, neurodata_type=pynwb.NWBContainer)
+def check_dataset_compression(nwb_container: pynwb.NWBContainer, bytes_threshold=2e6):
+    """
+    If the data in the TimeSeries object is a h5py.Dataset, check if it has compression enabled.
+
+    Will only run if the size of the h5py.Dataset is larger than bytes_threshold.
+    """
+    if (
+        "data" in nwb_container.__dict__["_AbstractContainer__field_values"]
+        and isinstance(nwb_container.data, h5py.Dataset)
+        and nwb_container.data.size * nwb_container.data.dtype.itemsize
+        > bytes_threshold
+        and nwb_container.data.compression is None
+    ):
+        return "Consider enabling compression when writing a large dataset."
```

Comment:
@CodyCBakerPhD , that's not what I meant (nor @oruebel I think). We want to check _all_ `h5py.Dataset` objects regardless of what they are named.

```suggestion
    for field in nwb_container.__fields__:
        obj = getattr(nwb_container, field)
        if (
            isinstance(obj, h5py.Dataset)
            and obj.size * obj.dtype.itemsize > bytes_threshold
            and obj.compression is None
        ):
        return f"Consider enabling compression when writing a large dataset {field}."
```

Reply:
Yes, checking all datasets seems the correct logic here. This is also an example where a check would need to be able to issue more than one warning as the check may end up finding multiple datasets within a single container that may benefit from compression and each of those should be treated as a separate item. To address this I would suggest to have checks return a tuple of ``NWBInspectorWarning`` type objects instead of a single string, where ``NWBInspectorWarning``would be a base class for warnings created by the NWBInspector so that you can also set the severity, color etc for each warning.

Reply:
yes, I agree this would benefit from the ability to return more than one warning per check, and we are likely to see that problem come up again. If we wanted to get really fancy, we could `yield` these warnings

Reply:
I'd like to keep the display logic separate though, so I'm not crazy about the idea of annotating with a color.

Reply:
I agree that the display logic should be separate. All I meant was that if you have checks return instances of some form of a warning class then you can more flexibly attach additional information to the warnings, e.g., to be able to filter warnings (I agree that color was a bad example as that should probably be handled elsewhere).

Reply:
Thanks for the clarification; I've generalized it to scan all the objects in the container now.

Note that `nwb_container.__fields__` returns empty (example: `test=pynwb.TimeSeries(name="test", unit="test", data=[1], rate=1.)`, `test.__fields__`); however, `test.fields` returns the dictionary, which is not iterated over in a safe manner.

Reply:
I'll have to play with the logic of the case of multiple `Dataset` objects in these fields however. Also, are there any cases where there might be a container within a container?

Reply:
>  which is not iterated over in a safe manner

what do you mean by this?

> Also, are there any cases where there might be a container within a container?

Yes, that happens all the time. NWBFile is a container. It's not a problem, since we are iterating over all containers, including containers that are inside other containers.

Reply:
> what do you mean by this?

Typo, sorry. If the `fields` attribute isn't present, or is empty, it fails safely.

> Yes, that happens all the time. NWBFile is a container. It's not a problem, since we are iterating over all containers, including containers that are inside other containers.

I see, the logic of `nwbfile.objects` (https://github.com/NeurodataWithoutBorders/pynwb/blob/e484bfd0208a9777bc6c45e44a4cfa3a900cffe3/src/pynwb/file.py#L448-L462) recursively unpacks these.